app/controllers/name_controller/create_and_edit_name.rb

```diff
@@ -291,7 +291,8 @@ def email_name_change_content
   # -------------
 
   def merge_names(new_name)
-    if in_admin_mode? || @name.mergeable? || new_name.mergeable?
+    if in_admin_mode? ||
+       ((@name.mergeable? || new_name.mergeable?) && !@name.dependency?)
```

Comment:
Hold on a second.  if @name is not mergeable but new_name is, it will reverse the order, in which case you should be checking if new_name.dependency? instead!

Actually, how is mergeable? different from dependency? ?  Maybe those two methods should be combined?

Reply:
Yes, I'd recommend having mergeable? call dependency?  And, if you wouldn't mind, would it be okay to expand the name "dependency?" a little to something like "depended_on_by_proposed_name?".  "Proposed?" doesn't quite work, I guess, because it's more complicated than that?  "Dependency?" just doesn't make it clear what it's checking that this name is dependent on... or is it the other way around, is it checking if something else is dependent on this name, and if so what?

Reply:
> Hold on a second. if @name is not mergeable but new_name is, it will reverse the order, in which case you should be checking if new_name.dependency? instead!
> 
> Actually, how is mergeable? different from dependency? ? Maybe those two methods should be combined?

Good questions. 
Short answer: `dependency?` is concerned with the literal name, while `mergeable?` is concerned with objects. I wish I had better terminology.

Ex: Start with two Names: 
>_Aaa bbb_, with `id == 123`
_Ccc ddd_, with `id == 456`

User edits _Aaa bbb_, => _Ccc ddd_

`dependency?` is concerned only with whether a (non-admin) user can delete the literal name _Aaa bbb_. That's why `dependency?` is called only on the object with that literal name --`@name`. If it's a dependency, then no merger regardless whether name 456 is mergeable.
(`mergeable?`, on the other hand is concerned only with whether a user can destroy an object.)

Does that make sense?
If you have better terminology, I'd love it. I was very confused by `perform_merger` for the longest time because "name" and variants could mean either the literal name or a Name object. (And maybe am still somewhat confused.)

Reply:
> Yes, I'd recommend having mergeable? call dependency? And, if you wouldn't mind, would it be okay to expand the name "dependency?" a little to something like "depended_on_by_proposed_name?". "Proposed?" doesn't quite work, I guess, because it's more complicated than that? "Dependency?" just doesn't make it clear what it's checking that this name is dependent on... or is it the other way around, is it checking if something else is dependent on this name, and if so what?

1. I don't think `mergeable?` should call `dependency?` for the reasons stated in https://github.com/MushroomObserver/mushroom-observer/pull/704#discussion_r525627409

2. You raise a legit concern about the method name `dependency?`. 
I struggled with that method name, and ended up borrowing `dependency` from gem, where "dependencies are what your gem needs to work." I'll try to come up with something more transparent.

Reply:
Yes, I see your point, I'm really struggling with this, too!

Hmmm...

Okay, now I'm wondering if this PR is going about this all wrong (forgive me!)  It's not the merge or delete we care about, per se, but the change to text_name.  Do you follow me?  That is, we should not allow a user to change text_name attribute of Agaricus even if it doesn't result in a merger / deletion.  If you change the genus name, all of the subtaxa must also change.

In fact, now that we're really getting to the heart of the matter, we shouldn't really allow someone to willy nilly change a name someone is tracking, either, even if it doesn't result in a merger.  ...

But no, I'm finally remembering... the point of Name#mergeable? is to prevent *irreversible changes*.  Your Name#dependency? is concerned with preventing *dangerous changes*.  I still think we should be checking this upon change of text_name, not when checking if a merge is acceptable, though.

So in summary, mergeable? checks if a merge can be undone.  If one of the two names is unused, then we can always recreate the name, no harm done.  And dependency? is checking whether there is a reference to the old text_name somewhere (either in a classification string or in another name's actual text_name).  Regardless of whether the change results in a merger, this will still invalidate the reference.

Would "would_merge_be_destructive?" be a better name for "mergeable?"?

And would... "external_references_to_text_name?" be a better name for "dependency?"?

And please have changes to text_name consult the latter (whatever you choose to call it), instead of merge.

Lastly, please consider having changes to the generic part of a species name also trigger this if the new name's genus doesn't already exist.  This is a bit tricky, though.  Use case: User tries to change "Amanita muscaria" to "Bogus othername".  The name "Bogus" doesn't exist.  Either this leaves the name without a parent or we have "Bogus" spring into existence automatically.  The former is dangerous.  The latter... is that reversible?  I guess so, just change the name back, then delete "Bogus".  I guess that's alright.  But would you mind making sure "Bogus" springs into existence if a user does that?  We can afford to have species without a genus hanging about.

You know, what we really need here is a better data structure.  The classification string is a mess.  I realize the reason for it is that Nathan doesn't want to lock us into any sort of strictly hierarchical classification structure.  But this is downright dangerous.  How many bogus names are present in various classification strings??  If we implemented it as a simple Name#parent_id attribute instead, changing the text_name of a name would not affect the classification hierarchy at all.  We would still have the genus name part of the text_name and other related attributes, so any change to a text_name would require propagating that change down to the name's descendants.  But arguably we should be doing that now, anyway.

I guess "broken" classification strings really just isn't a big deal.  So what?  Let our users keep things clean, right?  Gah, what a mess. :(  The important thing is just to prevent destruction of data.

And you aren't talking about preventing people from making dangerous changes, just having it notify us about them.  Then we can deal with them by hand as necessary.  ... But I still think the place to check dependency? is when text_name changes, not when the name is merged, no matter which of these many views you back!! :)

Reply:
Wow! Thanks for all those ideas!
(Just today I had explained to a group of friend that coding might seem super-boring to them, but there are always unexpected problems and complications, which is what makes it interesting and fun.)
I'll give this some thought and get back to you tomorrow. (I'm too tired now.)

Reply:
I'll work on all your suggestions (except revising the classification system). 
- There likely will be more attempts to change a bogus name to a real one (e.g., _British soldier_ to _Cladonia cristatella_) than vice versa. But that should not increase the need for admin intervention; intervention likely would have been already required because that change would require a merger.
- It will be challenging to avoid/reduce spaghetti in the controller. But it's also an opportunity to simplify things, e.g., by moving stuff to the model.
- I'm making this a Draft PR for now. I will let you know when I think it's again ready for merger. (Might take me a week.)

 a week.)

Reply:
PS: Today I will push changes responding to your other comments.
As I do so, I will "Resolve conversation" in Github, which will make it easier to see the remaining issues. 
(GitHub does this by hiding your Comment and any responses. To see them, click "Show resolved".)

Reply:
Yes. Genus _Ccc_ is automatically created when **editing** _Aaa bbb_ to _Ccc ddd_ if _Ccc_ does not already exist.
> Use case: User tries to change "Amanita muscaria" to "Bogus othername". The name "Bogus" doesn't exist. ... But would you mind making sure "Bogus" springs into existence if a user does that? We can't afford to have species without a genus hanging about.

Reply:
Does editing _Aaa bbb_ to _Ccc anything_ delete _Aaa_ if it no longer has subtaxa? ANSWER: NO 
For now, that's how it should be. But maybe create another ticket to delete _Aaa_ if it's not a dependency.